### ISSUE TITLE
TabController's index and animation is not working properly when a new tab is selected while there is a transition still occurring

### ISSUE THREAD

**Issue by lifeinacastle** [flutter/flutter#113020]
## Steps to Reproduce

1. Use the sample app navigation_and_routing.
2. In books.dart add a print statement in the method _handleTabIndexChanged to see what's the value of _tabController.indexIsChanging
3. In the app, start in the 'Popuplar' tab, Tap on the 'New' Tab and before it finishes quickly tap on the 'All' tab. Make the tabController animationDuration longer to see it easily

**Expected results:** 
The tab starts to navigate to the New tab, indexIsChanging is true. When it arrives to 'New' Tab indexIsChanging is false. Because the 'All' tab was pressed before arriving to 'New' tab, as soon as it arrives to 'New', it starts navigating to the 'All' tab and **indexIsChanging is true**.

**Actual results:** 
When it arrives to the 'New' tab and starts navigating to the 'All' tab, indexIsChanging never changes to true, it stays as false while the tabbar is transitioning. The same happens if you put a listener to the animation status.

**Video:** 
I put a setstate to show the value of indexIsChanging in the appbar.

https://user-images.githubusercontent.com/63162685/194283072-58def02d-524c-46b7-81c5-b8db62beb3db.mp4

**Comment by darshankawar** [comment]
@lifeinacastle 
Thanks for the report. Can you provide a [complete minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example), preferably in single file (main.dart) that we can directly use to verify this behavior ? Also provide `flutter doctor -v`.

**Comment by lifeinacastle** [comment]
```
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage>
    with SingleTickerProviderStateMixin {
  late TabController _tabController;
  bool transitioning = false;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(
        length: 3, vsync: this, animationDuration: Duration(seconds: 3))
      ..addListener(_handleTabIndexChanged);
  }

  void _handleTabIndexChanged() {
    setState(() {
      transitioning = _tabController.indexIsChanging;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('tr: $transitioning'),
        bottom: TabBar(
          controller: _tabController,
          tabs: const [
            Tab(
              text: 'Popular',
              icon: Icon(Icons.people),
            ),
            Tab(
              text: 'New',
              icon: Icon(Icons.new_releases),
            ),
            Tab(
              text: 'All',
              icon: Icon(Icons.list),
            ),
          ],
        ),
      ),
      body: TabBarView(
        controller: _tabController,
        children: const [
          ColoredBox(color: Colors.red),
          ColoredBox(color: Colors.green),
          ColoredBox(color: Colors.blue)
        ],
      ),
    );
  }
}

```

flutter doctor -v :

```
`[✓] Flutter (Channel stable, 3.0.3, on macOS 12.5.1 21G83 darwin-x64, locale en-GB)
    • Flutter version 3.0.3 at /Users/dvigara/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 676cefaaff (4 months ago), 2022-06-22 11:34:49 -0700
    • Engine revision ffe7b86a1e
    • Dart version 2.17.5
    • DevTools version 2.12.2

[!] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/dvigara/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.71.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (3 available)
    • GM1910 (mobile) • 888835cf • android-arm64  • Android 11 (API 30)
    • macOS (desktop) • macos    • darwin-x64     • macOS 12.5.1 21G83 darwin-x64
    • Chrome (web)    • chrome   • web-javascript • Google Chrome 105.0.5195.125

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.`
```

**Comment by darshankawar** [comment]
@lifeinacastle 
I tried the code sample using latest stable version (3.3.4) but didn't find any unexpected behavior.


https://user-images.githubusercontent.com/67046386/194551144-2177343b-d932-48b9-932b-85de82ef0f9a.mov

I see that you are on 3.0.3, can you upgrade to latest stable or master and try again to see if you still get same behavior ?

**Comment by lifeinacastle** [comment]
Hello, thank you for your quick response. 

There is no issue when navigating in the same way that you do in your video. But it happens during a transition when there is a tap to the third screen:

Navigate from the first tab to the second, and before it arrives to the second quickly tap the third tab. That's how I reproduce the issue. In the second 9 of my video you can see the quick taps.

**Comment by lifeinacastle** [comment]
I've upgraded to Flutter SDK 3.3.4 and I'm still facing the issue.

I've looked more carefully at your video, and the issue can be seen in the second 4 and 5. I've attached two images where you can see how the transition parameter goes from true to false but the transition hasn't finished. 

<img width="368" alt="transition_true" src="https://user-images.githubusercontent.com/63162685/194579212-7c1f03e3-bebf-44be-a5f0-c176a0e846a6.png">

<img width="368" alt="transition_false" src="https://user-images.githubusercontent.com/63162685/194579234-5bcdc4ff-c7f9-4f67-84b9-619869493b27.png">

**Comment by darshankawar** [comment]
I see. Thanks for the feedback. Using the exact steps to replicate as mentioned, I see the same behavior as reported.



> Navigate from the first tab to the second, and before it arrives to the second quickly tap the third tab. That's how I reproduce the issue. In the second 9 of my video you can see the quick taps.

<details>
<summary> stable, master flutter doctor -v </summary>

```

[✓] Flutter (Channel stable, 3.3.4, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.3.4 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision eb6d86ee27 (2 days ago), 2022-10-04 22:31:45 -0700
    • Engine revision c08d7d5efc
    • Dart version 2.18.2
    • DevTools version 2.15.0

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[✓] Flutter (Channel master, 3.5.0-4.0.pre, on macOS 12.2.1 21D62 darwin-x64,
    locale en-GB)
    • Flutter version 3.5.0-4.0.pre on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fe9b598c12 (12 hours ago), 2022-10-09 12:37:38 -0400
    • Engine revision ac1fec6367
    • Dart version 2.19.0 (build 2.19.0-295.0.dev)
    • DevTools version 2.18.0
    
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

<details>
<summary> runnable code sample </summary>

```

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage>
    with SingleTickerProviderStateMixin {
  late TabController _tabController;
  bool transitioning = false;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(
        length: 3, vsync: this, animationDuration: Duration(seconds: 3))
      ..addListener(_handleTabIndexChanged);
  }

  void _handleTabIndexChanged() {
    setState(() {
      transitioning = _tabController.indexIsChanging;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('tr: $transitioning'),
        bottom: TabBar(
          controller: _tabController,
          tabs: const [
            Tab(
              text: 'Popular',
              icon: Icon(Icons.people),
            ),
            Tab(
              text: 'New',
              icon: Icon(Icons.new_releases),
            ),
            Tab(
              text: 'All',
              icon: Icon(Icons.list),
            ),
          ],
        ),
      ),
      body: TabBarView(
        controller: _tabController,
        children: const [
          ColoredBox(color: Colors.red),
          ColoredBox(color: Colors.green),
          ColoredBox(color: Colors.blue)
        ],
      ),
    );
  }
}

```

</details>

**Comment by bleroux** [comment]
@lifeinacastle
Many thanks for this detailed report :+1: .
Looking at the repro, it seems that `TabController.indexIsChanging` is set to false at the exact moment when the `TabBar` indicator animation has ended, but at this moment the `TabBarView` is still animating which is confusing.

This is illustrated with the two screenshots your provided in https://github.com/flutter/flutter/issues/113020#issuecomment-1271680806. At second 4, the TabBar indicator is not yet at his final position. At second 5, it is but the TabBarView is still animating.

I have to investigate more to understand why `TabBar` indicator animation and `TabBarView` animation are not perfectly synchronised.

**Comment by bleroux** [comment]
The `TabBar` indicator and `TabBarView` page animations are desynchronized as illustrated in the video below:
- At second 1, I clicked on the second tab (both animations starts).
- At second 2, I clicked on the first tab (the TabBar indicator animation stopped and restarted from the current position, it reached its final position at second 4. The TabBarView page animation, from the first page to the second page, lasted until second 4 and then it animated to the first page, it ended at second 7).

[Capture vidéo du 2022-12-13 15-29-03.webm](https://user-images.githubusercontent.com/840911/207363824-1495aa88-d854-410c-8a5b-dbdd020cde2b.webm)

The root cause seems to be that `TabBarView` is awaiting for the animation to fully complete:
- https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/tabs.dart#L1596
- https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/tabs.dart#L1623

This change was introduced in https://github.com/flutter/flutter/pull/57574.

@HansMuller Is the expected feature is that both animations are synchronized? If yes, should both animations stop and restart when a new tab is selected?
